File: web/src/api/jvm.ts
import type { JvmSnapshot, DependencyEdge, DependencyCausesResponse } from '../types/jvm'
import { apiFetch } from './client'

export function fetchJVMServices(): Promise<string[]> {
  return apiFetch<string[]>('/api/v1/jvm/services')
}

export function fetchJVMHealth(service: string): Promise<JvmSnapshot> {
  return apiFetch<JvmSnapshot>(`/api/v1/jvm/health/${encodeURIComponent(service)}`)
}

export function fetchJVMHistory(service: string, windowMinutes = 60): Promise<JvmSnapshot[]> {
  return apiFetch<JvmSnapshot[]>(
    `/api/v1/jvm/history/${encodeURIComponent(service)}?window_minutes=${windowMinutes}`,
  )
}

export function fetchDependencyGraph(): Promise<DependencyEdge[]> {
  return apiFetch<DependencyEdge[]>('/api/v1/dependency/graph')
}

export function fetchDependencyCauses(service: string): Promise<DependencyCausesResponse> {
  return apiFetch<DependencyCausesResponse>(
    `/api/v1/dependency/${encodeURIComponent(service)}/causes`,
  )
}

File: internal/handler/jvm.go
package handler

import (
	"net/http"

	"github.com/go-chi/chi/v5"
)

// jvm.go proxies JVM analytics endpoints to javi-forecast (/api/jvm/*)
// and Granger causality dependency endpoints (/dependency/*).
// Both services share the same forecastBaseURL / proxyForecast helper
// defined in forecast.go.

// -----------------------------------------------------------------------
// JVM Analytics — GET /api/v1/jvm/*
// -----------------------------------------------------------------------

// GetJVMServices — GET /api/v1/jvm/services
// Proxies to javi-forecast GET /api/jvm/services
// Returns: JSON array of service name strings.
func GetJVMServices(w http.ResponseWriter, r *http.Request) {
	proxyForecast(w, r, "/api/jvm/services")
}

// GetJVMHealth — GET /api/v1/jvm/health/{service}
// Proxies to javi-forecast GET /api/jvm/health/{service_name}
// Returns: latest JvmMetricEvent snapshot for the service.
func GetJVMHealth(w http.ResponseWriter, r *http.Request) {
	name := chi.URLParam(r, "service")
	if name == "" {
		writeError(w, http.StatusBadRequest, "service name is required")
		return
	}
	proxyForecast(w, r, "/api/jvm/health/"+name)
}

// GetJVMHistory — GET /api/v1/jvm/history/{service}
// Proxies to javi-forecast GET /api/jvm/history/{service_name}
// Query params forwarded: ?window_minutes=60
func GetJVMHistory(w http.ResponseWriter, r *http.Request) {
	name := chi.URLParam(r, "service")
	if name == "" {
		writeError(w, http.StatusBadRequest, "service name is required")
		return
	}
	proxyForecast(w, r, "/api/jvm/history/"+name)
}

// -----------------------------------------------------------------------
// Granger Causality — GET /api/v1/dependency/*
// -----------------------------------------------------------------------

// GetDependencyGraph — GET /api/v1/dependency/graph
// Proxies to javi-forecast GET /dependency/graph
// Returns: all Granger-causality edges (source, target, p_value, max_lag).
func GetDependencyGraph(w http.ResponseWriter, r *http.Request) {
	proxyForecast(w, r, "/dependency/graph")
}

// GetDependencyCauses — GET /api/v1/dependency/{service}/causes
// Proxies to javi-forecast GET /dependency/{service}/causes
// Returns: root-cause services and upstream edges for the given service.
func GetDependencyCauses(w http.ResponseWriter, r *http.Request) {
	name := chi.URLParam(r, "service")
	if name == "" {
		writeError(w, http.StatusBadRequest, "service name is required")
		return
	}
	proxyForecast(w, r, "/dependency/"+name+"/causes")
}
